File: day20_FE_AT/EmpMan_phase2/tests/pages/AttendancePage.js
/**
 * Page Object Model – Attendance Module
 * Tabs: Daily Attendance, Pending Leave Requests
 * Features: KPI cards, search, dept filter, status filter, approve/reject leave.
 */
class AttendancePage {
  constructor(page) {
    this.page = page;

    this.container         = page.locator('.attendance');
    this.pageTitle         = page.locator('.dashboard-title').first();

    // Tabs
    this.attendanceTab     = page.locator('.tab-btn', { hasText: /Daily Attendance/i });
    this.leaveTab          = page.locator('.tab-btn', { hasText: /Leave Requests/i });

    // KPI cards
    this.kpiCards          = page.locator('.kpi-card');

    // Search / Filter bar
    this.searchInput       = page.locator('input[placeholder*="Search"]').first();
    this.deptFilter        = page.locator('select').first();
    this.statusFilter      = page.locator('select').nth(1);

    // Attendance roster table
    this.rosterTable       = page.locator('.data-table, table').first();
    this.rosterRows        = page.locator('tbody tr');

    // Leave request rows
    this.leaveRows         = page.locator('tbody tr');
    this.approveButtons    = page.locator('button', { hasText: /Approve/i });
    this.rejectButtons     = page.locator('button', { hasText: /Reject/i });
  }

  async waitForLoad() {
    await this.container.waitFor({ state: 'visible', timeout: 15000 });
    await this.page.waitForTimeout(800);
  }

  async clickAttendanceTab() {
    await this.attendanceTab.click();
    await this.page.waitForTimeout(500);
  }

  async clickLeaveTab() {
    await this.leaveTab.click();
    await this.page.waitForTimeout(500);
  }

  async search(term) {
    await this.searchInput.fill(term);
    await this.page.waitForTimeout(400);
  }

  async filterByDepartment(dept) {
    await this.deptFilter.selectOption(dept);
    await this.page.waitForTimeout(400);
  }

  async filterByStatus(status) {
    await this.statusFilter.selectOption(status);
    await this.page.waitForTimeout(400);
  }

  async getKpiCount() {
    return this.kpiCards.count();
  }

  async getRosterRowCount() {
    return this.rosterRows.count();
  }

  async getLeaveRowCount() {
    return this.leaveRows.count();
  }

  async approveFirstPendingLeave() {
    await this.approveButtons.first().click();
    await this.page.waitForTimeout(600);
  }

  async rejectFirstPendingLeave() {
    await this.rejectButtons.first().click();
    await this.page.waitForTimeout(600);
  }

  /** Read the date badge visible in the header */
  async getDateBadgeText() {
    return this.page.locator('.current-time-badge').innerText();
  }
}

module.exports = { AttendancePage };

File: day20_FE_AT/EmpMan_phase2/tests-java/src/main/java/com/ems/pages/AttendancePage.java
package com.ems.pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;

public class AttendancePage {
    private final Page page;

    public final Locator container;
    public final Locator attendanceTab;
    public final Locator leaveTab;
    public final Locator rosterRows;
    public final Locator searchInput;

    public AttendancePage(Page page) {
        this.page = page;
        this.container = page.locator(".attendance-module, .attendance-container");
        this.attendanceTab = page.locator("button").filter(new Locator.FilterOptions().setHasText("Daily Attendance"));
        this.leaveTab = page.locator("button").filter(new Locator.FilterOptions().setHasText("Leave Requests"));
        this.rosterRows = page.locator("tbody tr");
        this.searchInput = page.locator("input[placeholder*='Search']").first();
    }

    public void waitForLoad() {
        container.waitFor();
    }

    public void clickLeaveTab() {
        leaveTab.click();
        page.waitForTimeout(500);
    }
}
